google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/ReadRowsResumptionStrategy.java

```diff
@@ -0,0 +1,153 @@
+ */
+@InternalApi
+public class ReadRowsResumptionStrategy<RowT>
+    implements StreamResumptionStrategy<ReadRowsRequest, RowT> {
+  private final RowAdapter<RowT> rowAdapter;
+  private ByteString lastKey = ByteString.EMPTY;
+  private long numProcessed;
+
+  public ReadRowsResumptionStrategy(RowAdapter<RowT> rowAdapter) {
+    this.rowAdapter = rowAdapter;
+  }
+
+  @Override
+  public boolean canResume() {
+    return true;
+  }
+
+  @Override
+  public StreamResumptionStrategy<ReadRowsRequest, RowT> createNew() {
+    return new ReadRowsResumptionStrategy<>(rowAdapter);
+  }
+
+  @Override
+  public void onProgress(RowT response) {
+    // Last key can come from both the last processed row key and a synthetic row marker. The
+    // synthetic row marker is emitted when the server has read a lot of data that was filtered out.
+    // So it can trim the start of the scan, but does not contribute to the row limit.
+    lastKey = rowAdapter.getKey(response);
+    if (!rowAdapter.isScanMarkerRow(response)) {
+      // Only real rows count towards the rows limit.
+      numProcessed++;
+    }
+  }
+
+  @Override
+  public ReadRowsRequest getResumeRequest(ReadRowsRequest request) {
+    if (lastKey.isEmpty()) {
+      return request;
+    }
+    Builder builder = request.toBuilder();
+
+    if (request.getRowsLimit() > 0) {
+      Preconditions.checkState(
+          request.getRowsLimit() >= numProcessed,
+          "Detected too many responses for the current row limit during a retry.");
+      builder.setRowsLimit(request.getRowsLimit() - numProcessed);
+    }
+
+    // Reset rows to scan.
+    builder.clearRows();
+
+    RowSet.Builder newRowSet = RowSet.newBuilder();
+
+    // Special case: empty query implies full table scan
+    if (request.getRows().getRowKeysList().isEmpty()
+        && request.getRows().getRowRangesList().isEmpty()) {
+      newRowSet.addRowRanges(RowRange.newBuilder().setStartKeyOpen(lastKey).build());
+    } else {
+      for (ByteString key : request.getRows().getRowKeysList()) {
+        if (ByteStringComparator.INSTANCE.compare(key, lastKey) > 0) {
+          newRowSet.addRowKeys(key);
+        }
+      }
+
+      for (RowRange rowRange : request.getRows().getRowRangesList()) {
+        RowRange.Builder rangeBuilder = RowRange.newBuilder();
+
+        switch (rowRange.getEndKeyCase()) {
+          case END_KEY_CLOSED:
+            if (ByteStringComparator.INSTANCE.compare(rowRange.getEndKeyClosed(), lastKey) <= 0) {
+              continue;
+            } else {
+              rangeBuilder.setEndKeyClosed(rowRange.getEndKeyClosed());
+            }
+            break;
+          case END_KEY_OPEN:
+            if (ByteStringComparator.INSTANCE.compare(rowRange.getEndKeyOpen(), lastKey) <= 0) {
+              continue;
+            } else {
+              rangeBuilder.setEndKeyOpen(rowRange.getEndKeyOpen());
+            }
+            break;
+          case ENDKEY_NOT_SET:
+            rangeBuilder.clearEndKey();
```

Comment:
Break after this. Make sure we have a test case for this.
Retry full table scan has no endkey.

Reply:
done